File: app/src/main/java/h_mal/appttude/com/driver/driverProfileFragment.java
package h_mal.appttude.com.driver;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ValueEventListener;
import com.squareup.picasso.Picasso;

import h_mal.appttude.com.driver.Objects.DriverProfileObject;
import h_mal.appttude.com.driver.Objects.DriversLicenseObject;
import h_mal.appttude.com.driver.Objects.PrivateHireObject;

import static h_mal.appttude.com.driver.FirebaseClass.*;
import static h_mal.appttude.com.driver.MainActivity.auth;
import static h_mal.appttude.com.driver.MainActivity.executeFragment;
import static h_mal.appttude.com.driver.MainActivity.fragmentManager;
import static h_mal.appttude.com.driver.MainActivity.getDateStamp;
import static h_mal.appttude.com.driver.MainActivity.mDatabase;


public class driverProfileFragment extends Fragment {

    private String TAG = this.getClass().getSimpleName();

    ImageView driverPic;
    TextView addPic;
    EditText forenames;
    EditText address;
    EditText postcode;
    EditText dob;
    EditText ni;
    TextView privateHire;
    TextView retrievedPH;
    EditText dateFirst;
    TextView driverLi;
    TextView retrievedDl;
    Button submit_driver;

    Uri filePath;
    Uri picUri;

    DatabaseReference driverProfileReference;

    PrivateHireObject privateHireObject;
    DriverProfileObject driverProfileObject;
    DriversLicenseObject driversLicenseObject;


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        driverProfileReference = mDatabase.child(USER_FIREBASE).child(auth.getCurrentUser().getUid())
                .child(DRIVER_FIREBASE);

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_driver_profile, container, false);

        driverPic = view.findViewById(R.id.driver_pic);
        addPic = view.findViewById(R.id.add_driver_pic);
        forenames = view.findViewById(R.id.names);
        address = view.findViewById(R.id.address);
        postcode  = view.findViewById(R.id.postcode);
        dob = view.findViewById(R.id.dob);
        ni = view.findViewById(R.id.ni_number);
        privateHire = view.findViewById(R.id.private_hire);
        retrievedPH = view.findViewById(R.id.retrievedPH);
        dateFirst = view.findViewById(R.id.date_first);
        driverLi = view.findViewById(R.id.drivers_li);
        retrievedDl = view.findViewById(R.id.retrieved_dl);
        submit_driver = view.findViewById(R.id.submit_driver);

        driverPic.setVisibility(View.GONE);

        driverProfileReference.addListenerForSingleValueEvent(valueEventListener);

        addPic.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ImageSelectorDialog imageSelectorDialog = new ImageSelectorDialog(getContext());
                imageSelectorDialog.setImageName("driver_pic"+getDateStamp());
                imageSelectorDialog.show();
            }
        });

        privateHire.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                executeFragment(new PrivateHireLicenseFragment());
            }
        });

        driverLi.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                executeFragment(new DriverLicenseFragment());
            }
        });

        dob.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DateDialog dateDialog = new DateDialog(getContext());
                dateDialog.init(dob);
            }
        });

        dateFirst.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DateDialog dateDialog = new DateDialog(getContext());
                dateDialog.init(dateFirst);
            }
        });

        submit_driver.setOnClickListener(submitOnClickListener);


        return view;
    }

    ValueEventListener valueEventListener = new ValueEventListener(){

        @Override
        public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
            try{
                privateHireObject = dataSnapshot.child(PRIVATE_HIRE_FIREBASE).getValue(PrivateHireObject.class);
            }catch (Exception e){
                Log.e(TAG, "onDataChange: ", e);
            }finally {
                if (privateHireObject != null){

                    retrievedPH.setText("Private Hire License no.: " +privateHireObject.getPhNumber()
                            +"\nPrivate Hire License expiry: " + privateHireObject.getPhExpiry());
                }
            }

            try {
                driverProfileObject = dataSnapshot.child(DRIVER_DETAILS_FIREBASE).getValue(DriverProfileObject.class);
            }catch (Exception e){
                Log.e(TAG, "onDataChange: ", e);
            }finally {
                if (driverProfileObject != null){
                    forenames.setText(driverProfileObject.getForenames());
                    address.setText(driverProfileObject.getAddress());
                    postcode.setText(driverProfileObject.getPostcode());
                    dob.setText(driverProfileObject.getDob());
                    dateFirst.setText(driverProfileObject.getDateFirst());
                    ni.setText(driverProfileObject.getNi());
                    driverPic.setVisibility(View.VISIBLE);
                    Picasso.get().load(driverProfileObject.getDriverPic()).into(driverPic);
                    picUri = Uri.parse(driverProfileObject.getDriverPic());

                }else {
                    driverPic.setVisibility(View.GONE);
                }
            }

            try{
                driversLicenseObject = dataSnapshot.child(DRIVERS_LICENSE_FIREBASE).getValue(DriversLicenseObject.class);
            }catch (Exception e){
                Log.e(TAG, "onDataChange: ", e);
            }finally {
                if (driversLicenseObject != null){
                    retrievedDl.setText("Driving License no.: " +driversLicenseObject.getLicenseNumber()
                            +"\nDriving License expiry: " + driversLicenseObject.getLicenseExpiry());
                }
            }
        }

        @Override
        public void onCancelled(@NonNull DatabaseError databaseError) {

        }
    };

    View.OnClickListener submitOnClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            String driverForename = forenames.getText().toString().trim();
            String AddressString = address.getText().toString().trim();
            String postCodeString = postcode.getText().toString().trim();
            String dobString = dob.getText().toString().trim();
            String niString = ni.getText().toString().trim();
            String dateFirstString = dateFirst.getText().toString().trim();

            if (    !TextUtils.isEmpty(driverForename) &&
                    !TextUtils.isEmpty(AddressString) &&
                    !TextUtils.isEmpty(postCodeString) &&
                    !TextUtils.isEmpty(dobString) &&
                    !TextUtils.isEmpty(niString) &&
                    !TextUtils.isEmpty(dateFirstString)){
                if (filePath != null){
                    Log.i(TAG, "onClick: new Image uploaded");
                    new FirebaseClass(getContext(),filePath,new FirebaseClass.Response(){
                        @Override
                        public void processFinish(Uri output) {
                            Log.i(TAG, "processFinish: ");
                            if (output != null){
                                picUri = output;
                                writeDriverToDb();
                                fragmentManager.popBackStack();
                            }

                        }
                    }).uploadImage(DRIVERS_LICENSE_FIREBASE,DRIVERS_LICENSE_FIREBASE + getDateStamp());
                }else{
                    Log.i(TAG, "onClick: pushing with same image");
                    writeDriverToDb();
                }

            }else {
                if (TextUtils.isEmpty(driverForename)){
                    forenames.setError("Field required");
                }
                if (TextUtils.isEmpty(AddressString)){
                    address.setError("Field required");
                }
                if (TextUtils.isEmpty(postCodeString)){
                    postcode.setError("Field required");
                }
                if (TextUtils.isEmpty(dobString)){
                    dob.setError("Field required");
                }
                if (TextUtils.isEmpty(niString)){
                    ni.setError("Field required");
                }
                if (TextUtils.isEmpty(dateFirstString)){
                    dateFirst.setError("Field required");
                }
                if (picUri == null){
                    Toast.makeText(getContext(), getString(R.string.image_required), Toast.LENGTH_SHORT).show();
                }
            }

        }

    };

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        new ImageSelectorResults(new ImageSelectorResults.FilepathResponse() {
            @Override
            public void processFinish(Uri output) {
                filePath = output;
            }
        }).Results(getActivity(),requestCode, resultCode, data,
                filePath,driverPic);
        driverPic.setVisibility(View.VISIBLE);

    }

    private void writeDriverToDb(){

        String forenameText = forenames.getText().toString().trim();
        String addressText = address.getText().toString().trim();
        String postcodeText = postcode.getText().toString().trim();
        String dobText = dob.getText().toString().trim();
        String niText = ni.getText().toString().trim();
        String datefirstText = dateFirst.getText().toString().trim();

        DriverProfileObject driverProfileObject = new DriverProfileObject(picUri.toString(),forenameText,
                addressText,postcodeText,dobText,niText,datefirstText);

        driverProfileReference.child(DRIVER_DETAILS_FIREBASE).setValue(driverProfileObject);
    }
}

File: app/src/main/java/h_mal/appttude/com/driver/Driver/DriverProfileFragment.kt
package h_mal.appttude.com.driver.Driver

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.fragment.app.Fragment
import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.OnFailureListener
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.UserProfileChangeRequest
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.ValueEventListener
import com.squareup.picasso.Picasso
import h_mal.appttude.com.driver.Global.DateDialog
import h_mal.appttude.com.driver.Global.FirebaseClass
import h_mal.appttude.com.driver.Global.ImageSelectorDialog
import h_mal.appttude.com.driver.Global.ImageSelectorResults
import h_mal.appttude.com.driver.Global.ImageSelectorResults.FilepathResponse
import h_mal.appttude.com.driver.MainActivity
import h_mal.appttude.com.driver.Objects.DriverProfileObject
import h_mal.appttude.com.driver.Objects.UserObject
import h_mal.appttude.com.driver.R


class DriverProfileFragment : Fragment() {
    private val TAG: String = this.javaClass.simpleName
    var driverPic: ImageView? = null
    var addPic: TextView? = null
    var forenames: EditText? = null
    var address: EditText? = null
    var postcode: EditText? = null
    var dob: EditText? = null
    var ni: EditText? = null
    var dateFirst: EditText? = null
    var submit_driver: Button? = null
    var filePath: Uri? = null
    var picUri: Uri? = null
    var driverProfileReference: DatabaseReference? = null
    var UID: String? = null
    var driverProfileObject: DriverProfileObject? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            UID = arguments!!.getString("user_id")
        } else {
            UID = MainActivity.auth!!.currentUser!!.uid
        }
        driverProfileReference =
            MainActivity.mDatabase!!.child(FirebaseClass.USER_FIREBASE).child(
                (UID)!!
            )
                .child(FirebaseClass.DRIVER_FIREBASE)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        val view: View = inflater.inflate(R.layout.fragment_driver_profile, container, false)
        driverPic = view.findViewById(R.id.driver_pic)
        addPic = view.findViewById(R.id.add_driver_pic)
        forenames = view.findViewById(R.id.names)
        address = view.findViewById(R.id.address)
        postcode = view.findViewById(R.id.postcode)
        dob = view.findViewById(R.id.dob)
        ni = view.findViewById(R.id.ni_number)
        dateFirst = view.findViewById(R.id.date_first)
        submit_driver = view.findViewById(R.id.submit_driver)
        MainActivity.viewController!!.progress(View.VISIBLE)
        driverProfileReference!!.addListenerForSingleValueEvent(valueEventListener)
        addPic.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View) {
                val imageSelectorDialog: ImageSelectorDialog = ImageSelectorDialog((context)!!)
                imageSelectorDialog.setImageName("driver_pic" + MainActivity.Companion.dateStamp)
                imageSelectorDialog.show()
            }
        })
        dob.setOnClickListener(View.OnClickListener {
            val dateDialog: DateDialog = DateDialog((context)!!)
            dateDialog.init(dob)
        })
        dateFirst.setOnClickListener(View.OnClickListener {
            val dateDialog: DateDialog = DateDialog((context)!!)
            dateDialog.init(dateFirst)
        })
        submit_driver.setOnClickListener(submitOnClickListener)
        return view
    }

    var valueEventListener: ValueEventListener = object : ValueEventListener {
        override fun onDataChange(dataSnapshot: DataSnapshot) {
            MainActivity.viewController!!.progress(View.GONE)
            try {
                driverProfileObject =
                    dataSnapshot.child(FirebaseClass.DRIVER_DETAILS_FIREBASE).getValue(
                        DriverProfileObject::class.java
                    )
            } catch (e: Exception) {
                Log.e(TAG, "onDataChange: ", e)
            } finally {
                if (driverProfileObject != null) {
                    driverProfileObject?.apply {
                        forenames.setText(forenames)
                        address.setText(address)
                        postcode.setText(postcode)
                        dob.setText(dob)
                        dateFirst.setText(dateFirst)
                        ni.setText(ni)
                        Picasso.get().load(driverPic)
                            .into(MainActivity.loadImage(driverPic))
                        picUri = Uri.parse(driverPic)
                    }

                }
            }
        }

        override fun onCancelled(databaseError: DatabaseError) {
            MainActivity.viewController!!.progress(View.GONE)
        }
    }
    var submitOnClickListener: View.OnClickListener = View.OnClickListener {
        val driverForename: String = forenames!!.text.toString().trim { it <= ' ' }
        val AddressString: String = address!!.text.toString().trim { it <= ' ' }
        val postCodeString: String = postcode!!.text.toString().trim { it <= ' ' }
        val dobString: String = dob!!.text.toString().trim { it <= ' ' }
        val niString: String = ni!!.text.toString().trim { it <= ' ' }
        val dateFirstString: String = dateFirst!!.text.toString().trim { it <= ' ' }
        if ((!TextUtils.isEmpty(driverForename) &&
                    !TextUtils.isEmpty(AddressString) &&
                    !TextUtils.isEmpty(postCodeString) &&
                    !TextUtils.isEmpty(dobString) &&
                    !TextUtils.isEmpty(niString) &&
                    !TextUtils.isEmpty(dateFirstString))
        ) {
            if (filePath == null && picUri == null) {
                Toast.makeText(context, "No Driver image", Toast.LENGTH_SHORT).show()
                MainActivity.viewController!!.progress(View.GONE)
            } else {
                MainActivity.viewController!!.progress(View.VISIBLE)
                if (filePath != null) {
                    FirebaseClass(context, filePath, object : FirebaseClass.Response {
                        override fun processFinish(output: Uri?) {
                            Log.i(TAG, "processFinish: ")
                            if (output != null) {
                                picUri = output
                                writeDriverToDb()
                            } else {
                                MainActivity.viewController!!.progress(View.GONE)
                            }
                        }
                    }).uploadImage(
                        FirebaseClass.DRIVERS_LICENSE_FIREBASE,
                        FirebaseClass.DRIVERS_LICENSE_FIREBASE + MainActivity.Companion.dateStamp
                    )
                } else {
                    Log.i(TAG, "onClick: pushing with same image")
                    writeDriverToDb()
                }
            }
        } else {
            if (TextUtils.isEmpty(driverForename)) {
                forenames!!.error = "Field required"
            }
            if (TextUtils.isEmpty(AddressString)) {
                address!!.error = "Field required"
            }
            if (TextUtils.isEmpty(postCodeString)) {
                postcode!!.error = "Field required"
            }
            if (TextUtils.isEmpty(dobString)) {
                dob!!.error = "Field required"
            }
            if (TextUtils.isEmpty(niString)) {
                ni!!.error = "Field required"
            }
            if (TextUtils.isEmpty(dateFirstString)) {
                dateFirst!!.error = "Field required"
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)
        ImageSelectorResults().Results(
            activity, requestCode, resultCode, data,
            filePath, driverPic, object : FilepathResponse {
                override fun processFinish(output: Uri?) {
                    filePath = output
                }
            })
    }

    private fun writeDriverToDb() {
        val forenameText: String = forenames!!.text.toString().trim({ it <= ' ' })
        val addressText: String = address!!.text.toString().trim({ it <= ' ' })
        val postcodeText: String = postcode!!.text.toString().trim({ it <= ' ' })
        val dobText: String = dob!!.text.toString().trim({ it <= ' ' })
        val niText: String = ni!!.text.toString().trim({ it <= ' ' })
        val datefirstText: String = dateFirst!!.text.toString().trim({ it <= ' ' })
        val driverProfileObject: DriverProfileObject = DriverProfileObject(
            picUri.toString(), forenameText,
            addressText, postcodeText, dobText, niText, datefirstText
        )
        if ((UID == MainActivity.auth!!.currentUser!!.uid)) {
            val profileUpdatesBuilder: UserProfileChangeRequest.Builder =
                UserProfileChangeRequest.Builder()
            profileUpdatesBuilder.setPhotoUri(picUri)
            val profileUpdates: UserProfileChangeRequest = profileUpdatesBuilder.build()
            MainActivity.auth!!.currentUser!!.updateProfile(profileUpdates)
                .addOnCompleteListener(object : OnCompleteListener<Void?> {
                    override fun onComplete(task: Task<Void?>) {
                        if (task.isSuccessful) {
                            Log.d(TAG, "User profile updated.")
                            MainActivity.viewController!!.reloadDrawer()
                            MainActivity.mDatabase!!.child(FirebaseClass.USER_FIREBASE)
                                .child(
                                    MainActivity.auth!!.currentUser!!.uid
                                ).child("user_details")
                                .setValue(
                                    UserObject(
                                        MainActivity.auth!!.currentUser!!
                                            .displayName,
                                        MainActivity.auth!!.currentUser!!
                                            .email,
                                        picUri.toString()
                                    )
                                )
                        }
                    }
                })
                .addOnFailureListener(object : OnFailureListener {
                    override fun onFailure(e: Exception) {
                        Log.e(TAG, "onFailure: ", e)
                    }
                })
        }
        driverProfileReference!!.child(FirebaseClass.DRIVER_DETAILS_FIREBASE)
            .setValue(driverProfileObject)
            .addOnCompleteListener(object : OnCompleteListener<Void?> {
                override fun onComplete(task: Task<Void?>) {
                    if (task.isSuccessful) {
                        MainActivity.approvalsClass!!.setStatusCode(
                            UID,
                            FirebaseClass.DRIVER_DETAILS_FIREBASE + FirebaseClass.APPROVAL_CONSTANT,
                            FirebaseClass.APPROVAL_PENDING
                        )
                    }
                    MainActivity.viewController!!.progress(View.GONE)
                    MainActivity.fragmentManager!!.popBackStack()
                }
            })
    }
}